lib/Agrammon/DB/Dataset.pm6

```diff
@@ -542,4 +507,79 @@ class Agrammon::DB::Dataset does Agrammon::DB {
         }
     }
 
+    method store-branch-data(@vars, Str $instance, %options, @fractions) {
+        my $dataset-name = $!name;
+
+        # pg-array syntax: "{1,2,3}"
+        my $fractions-pg = '{' ~ @fractions[*;*].join(',') ~ '}';
+
+        my @branch-variables;
+        # Get variable ids and names
+        self.with-db: -> $db {
+            my $username = $!user.username;
+            @branch-variables = $db.query(q:to/SQL/, $username, $dataset-name, |@vars, $instance).hashes;
+            SELECT data_id, data_var
+                  FROM data_new
+                 WHERE data_dataset=dataset_name2id($1,$2)
+                   AND data_var IN ($3,$4)
+                   AND data_instance = $5
+                   ORDER BY data_id
+            SQL
+        }
+
+        for @branch-variables -> %var {
+            my $var-id   = %var<data_id>;
+            my $var-name = %var<data_var>;
+
+            # pg array syntax: { "test_1", "test_2" }
+            my $options-pg = '{' ~ %options{$var-name}.map(*.subst(' ', '_', :g) ).join(',') ~ '}';
+
+            self.with-db: -> $db {
+                my $ret = $db.query(q:to/SQL/, $var-id, $fractions-pg, $options-pg);
+                    INSERT INTO branches (branches_var, branches_data, branches_options)
+                                  VALUES ($1, $2, $3)
+                    ON CONFLICT (branches_var)
+                    DO
+                        UPDATE SET branches_data = EXCLUDED.branches_data,
+                                   branches_options = EXCLUDED.branches_options
+                    SQL
+                die "Couldn't save branch data" unless $ret;
+
+                $db.query(q:to/SQL/, $!user.username, $dataset-name);
+                    UPDATE dataset SET dataset_mod_date = CURRENT_TIMESTAMP
+                     WHERE dataset_id=dataset_name2id($1,$2)
+                SQL
+            }
+
+            return 'Branch data saved';
+        }
+    }
+
+    method load-branch-data(@var-names, $instance) {
+        my $data;
+        self.with-db: -> $db {
+            my $username = $!user.username;
+            my @vars = $db.query(q:to/SQL/, $username, $!name, |@var-names, $instance).arrays;
+                SELECT data_id
+                  FROM data_new
+                 WHERE data_dataset=dataset_name2id($1,$2)
+                   AND data_var IN ($3,$4)
+                   AND data_instance = $5
+                 ORDER BY data_id
+            SQL
+
+            my $branches = $db.query(q:to/SQL/, |@vars[*;*]).hashes;
+                SELECT branches_data, branches_options
+                  FROM branches
+                 WHERE branches_var in ($1,$2)
+                 ORDER BY branches_id
+            SQL
+            $data = {
+                fractions => $branches[0]<branches_data>,
+                options   => [ $branches[0]<branches_options>, $branches[1]<branches_options> ]
```

Comment:
Seems here we *do* rely on `DB::Pg` doing the right thing with the arrays.

Reply:
I knew that `SELECT` returns them as arrays. Should have guessed it would work on `INSERT` as well.